File: src/main/java/me/nepnep/nepbot/message/command/commands/info/WhoIs.java
package me.nepnep.nepbot.message.command.commands.info;

import me.nepnep.nepbot.message.command.Category;
import me.nepnep.nepbot.message.command.ICommand;
import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.Permission;
import net.dv8tion.jda.api.entities.Member;
import net.dv8tion.jda.api.entities.Role;
import net.dv8tion.jda.api.entities.TextChannel;
import net.dv8tion.jda.api.entities.User;
import net.dv8tion.jda.api.events.message.guild.GuildMessageReceivedEvent;
import net.dv8tion.jda.api.exceptions.ErrorResponseException;

import java.time.LocalDate;
import java.time.OffsetDateTime;
import java.time.temporal.ChronoUnit;
import java.util.EnumSet;
import java.util.List;
import java.util.stream.Collectors;

public class WhoIs implements ICommand {
    @Override
    public void execute(List<String> args, GuildMessageReceivedEvent event) {
        List<Member> mentioned = event.getMessage().getMentionedMembers();

        if (mentioned.size() >= 1) {
            sendEmbed(mentioned.get(0).getUser(), event);
        } else if (args.size() == 0) {
            sendEmbed(event.getAuthor(), event);
        } else {
            try {
                sendEmbed(event.getJDA().retrieveUserById(Long.parseLong(args.get(0))).complete(), event);
            } catch (NumberFormatException | NullPointerException | ErrorResponseException e) {
                event.getChannel().sendMessage("Invalid id!").queue();
            }
        }
    }
    private void sendEmbed(User user, GuildMessageReceivedEvent event) {
        EmbedBuilder builder = new EmbedBuilder();
        LocalDate now = LocalDate.now();
        TextChannel channel = event.getChannel();
        String id = user.getId();
        String tag = user.getAsTag();
        String avatarUrl = user.getAvatarUrl();

        OffsetDateTime timeCreated = user.getTimeCreated();
        LocalDate createdDate = timeCreated.toLocalDate();
        long sinceCreated = ChronoUnit.DAYS.between(createdDate, now);
        try {
            Member member = event.getGuild().retrieveMember(user).complete();

            OffsetDateTime timeJoined = member.getTimeJoined();
            LocalDate joinedDate = timeJoined.toLocalDate();
            long sinceJoined = ChronoUnit.DAYS.between(joinedDate, now);

            List<Role> roles = member.getRoles();
            List<String> rolesString = roles.stream()
                    .map(Role::getName)
                    .collect(Collectors.toList());
            EnumSet<Permission> permissions = member.getPermissions();
            List<String> permissionsString = permissions.stream()
                    .map(Permission::getName)
                    .collect(Collectors.toList());

            builder.setTitle(String.format("%s/%s", id, tag));

            builder.setThumbnail(avatarUrl);

            builder.addField("Created on: ", String.format("%s-%s-%s", timeCreated.getDayOfMonth(), timeCreated.getMonthValue(), timeCreated.getYear()), false);

            builder.addField("Days since created: ", String.valueOf(sinceCreated), false);

            builder.addField("Joined on: ", String.format("%s-%s-%s", timeJoined.getDayOfMonth(), timeJoined.getMonthValue(), timeJoined.getYear()), false);

            builder.addField("Days since joined: ", String.valueOf(sinceJoined), false);

            builder.addField("Roles: ", String.valueOf(rolesString), false);

            builder.addField("Permissions: ", String.valueOf(permissionsString), false);

            channel.sendMessage(builder.build()).queue();
        } catch (ErrorResponseException e) {
            builder.setTitle(String.format("%s/%s", id, tag));

            builder.setThumbnail(avatarUrl);

            builder.addField("Created on: ", String.format("%s-%s-%s", timeCreated.getDayOfMonth(), timeCreated.getMonthValue(), timeCreated.getYear()), false);

            builder.addField("Days since created: ", String.valueOf(sinceCreated), false);

            builder.addField("Not in guild", ":x:", false);

            channel.sendMessage(builder.build()).queue();
        }
    }
    @Override
    public String getInvoke() {
        return "whois";
    }
    @Override
    public Permission requiredPermission() {
        return null;
    }
    @Override
    public Category getCategory() {
        return Category.INFO;
    }
    @Override
    public String getDescription() {
        return "Gets a user's information: ;whois <Mention member> | <long id> | null";
    }
}

File: src/main/kotlin/me/nepnep/nepbot/message/command/commands/info/WhoIs.kt
package me.nepnep.nepbot.message.command.commands.info

import dev.minn.jda.ktx.coroutines.await
import dev.minn.jda.ktx.messages.EmbedBuilder
import me.nepnep.nepbot.message.command.AbstractCommand
import me.nepnep.nepbot.message.command.Category
import net.dv8tion.jda.api.entities.GuildMessageChannel
import net.dv8tion.jda.api.entities.User
import net.dv8tion.jda.api.events.message.MessageReceivedEvent
import net.dv8tion.jda.api.exceptions.ErrorResponseException
import java.time.LocalDate
import java.time.temporal.ChronoUnit

class WhoIs : AbstractCommand(
    "whois",
    Category.INFO,
    "Gets a user's information: ;whois <Mention member> | <long id> | null"
) {
    override suspend fun execute(args: List<String>, event: MessageReceivedEvent, channel: GuildMessageChannel) {
        val mentioned = event.message.mentions.members
        if (mentioned.isNotEmpty()) {
            sendEmbed(mentioned[0].user, event)
            return
        }

        if (args.isNotEmpty()) {
            try {
                val user = event.jda.retrieveUserById(args[0]).await()
                sendEmbed(user, event)
            } catch (e: RuntimeException) { //  ErrorResponseException | NumberFormatException
                channel.sendMessage("Invalid id!").queue()
            }
            return
        }
        sendEmbed(event.author, event)
    }

    private suspend fun sendEmbed(user: User, event: MessageReceivedEvent) {
        val channel = event.channel
        val now = LocalDate.now()

        val id = user.id
        val tag = user.asTag
        val avatarUrl = user.effectiveAvatarUrl

        val timeCreated = user.timeCreated
        val sinceCreated = ChronoUnit.DAYS.between(timeCreated.toLocalDate(), now)

        val builder = EmbedBuilder {
            title = "$id/$tag"
            thumbnail = avatarUrl
            field("Created on: ", "${timeCreated.dayOfMonth}-${timeCreated.monthValue}-${timeCreated.year}", false)
            field("Days since created: ", sinceCreated.toString(), false)
        }
        try {
            val member = event.guild.retrieveMember(user).await()
            val timeJoined = member.timeJoined
            val sinceJoined = ChronoUnit.DAYS.between(timeJoined.toLocalDate(), now)

            val roles = member.roles.map { it.name }
            val permissions = member.permissions.map { it.name }

            val embed = builder.apply { 
                field("Joined on: ", "${timeJoined.dayOfMonth}-${timeJoined.monthValue}-${timeJoined.year}", false)
                field("Days since joined: ", sinceJoined.toString(), false)
                field("Roles: ", roles.toString(), false)
                field("Permissions: ", permissions.toString(), false)
            }.build()
            if (embed.isSendable) {
                channel.sendMessageEmbeds(embed).queue()
                return
            }
            channel.sendMessage("Embed is too large!").queue()
        } catch (e: ErrorResponseException) {
            channel.sendMessageEmbeds(builder.apply { field("Not in guild", ":x:", false) }.build()).queue()
        }
    }
}
